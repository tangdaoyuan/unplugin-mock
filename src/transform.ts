import type { IncomingMessage, ServerResponse } from 'http'
import type { TransformResult } from 'unplugin'
import type { Options } from './types'

export function transform(_code: string, _id: string, _options: Options): TransformResult {
  return ''
}

export function transformRequest(
  _req: IncomingMessage,
  _res: ServerResponse,
  _next: (err?: any) => void,
): void {
  // eslint-disable-next-line no-console
  console.log(_req.url, _req.method)
  // TODO
  // filter url and method from request, mapping to mock server
  _next()
}
