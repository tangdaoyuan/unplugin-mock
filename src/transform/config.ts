import path from 'path'
import _module from 'node:module'
import fg from 'fast-glob'
import { buildSync } from 'esbuild'
import type { ExposeNodeModule, GeneralOptions } from '../types'

export function transformConfig(
  _options: GeneralOptions,
): any[] {
  // 1. get mock filepath from options
  const { mockPath } = _options
  if (!mockPath)
    return []

  // get absolute path
  const mockFiles = fg.sync(`${mockPath}/**/*.ts`)

  // 2. compile mock file to cjs (esm not supported to compiled by source code)
  const mockReqData = mockFiles.flatMap((mockFile) => {
    const mockFilePath = path.resolve(mockFile)
    const rawCjsCode = buildCjsFile(mockFilePath)
    // 3. import or require mock file in current context
    const mod = new _module(mockFilePath) as ExposeNodeModule
    mod.paths = module.paths
    mod.filename = mockFile
    mod._compile(rawCjsCode, mockFile)

    return mod.exports.default || []
  })

  // 4. return config
  return mockReqData
}

export function buildCjsFile(filePath: string) {
  if (!path.isAbsolute(filePath))
    filePath = path.resolve(filePath)
  const result = buildSync({
    entryPoints: [filePath],
    format: 'cjs',
    bundle: true,
    platform: 'node',
    write: false,
  })
  return result.outputFiles?.[0].text || ''
}
