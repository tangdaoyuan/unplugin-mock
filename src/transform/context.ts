import colors from 'picocolors'
import { parse } from 'regexparam'
import logger from '../logger'
import type { ModuleMockHandler } from '@/types'

export const QUICK_MOCK_DATA_KEY = 'quickReqData'
export const REGEX_MOCK_DATA_KEY = 'restReqData'

const quickHandlerStore = new Map<string, ModuleMockHandler>()
const regexHandlerStore = new Array<ModuleMockHandler>()

export function getMockHandler(url?: string, method?: string) {
  if (!url || !method)
    return null

  const handler = quickHandlerStore.get(`${method.toLowerCase()} ${url}`)
  if (handler)
    return handler

  for (const handler of regexHandlerStore) {
    const { url: _url, method: _method } = handler
    const matcher = parse(_url)
    if ((matcher.pattern.test(url) || handler.url === url)
      && _method.toLowerCase() === method.toLowerCase()
    )
      return handler
  }
  return null
}

export function setMockHandlerContext(values: ModuleMockHandler[], reset = true) {
  const logConflict = (source: ModuleMockHandler, target: ModuleMockHandler) => {
    logger.warn(`${colors.yellow('handler conflict as follows:')}
      ${colors.dim(colors.green(`${source.method.toUpperCase()} ${source.url}`))} ${colors.yellow(source._file)}
      ${colors.dim(colors.green(`${target.method.toUpperCase()} ${target.url}`))} ${colors.yellow(target._file)}`)
  }

  if (reset) {
    regexHandlerStore.splice(0, regexHandlerStore.length)
    quickHandlerStore.clear()
  }

  // restful priority > quick priority
  values
    .filter(handler => parse(handler.url).keys.length)
    .forEach((handler) => {
      const h = getMockHandler(handler.url, handler.method)
      if (h) {
        logConflict(handler, h)
        return
      }
      regexHandlerStore.push(handler)
    })
  values
    .filter(handler => !parse(handler.url).keys.length)
    .forEach((handler) => {
      const h = getMockHandler(handler.url, handler.method)
      if (h) {
        logConflict(handler, h)
        return
      }
      const key = `${handler.method.toLowerCase()} ${handler.url}`
      quickHandlerStore.set(key, handler)
    })
}
