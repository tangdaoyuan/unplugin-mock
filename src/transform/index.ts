import type { TransformResult } from 'unplugin'
import type { Options } from '../types'

export function transform(_code: string, _id: string, _options: Options): TransformResult {
  return ''
}
