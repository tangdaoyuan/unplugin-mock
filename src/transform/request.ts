import type { IncomingMessage, ServerResponse } from 'http'
import type { ResolvedConfig } from 'vite'
import { parse } from 'regexparam'
import colors from 'picocolors'
import logger from '../logger'
import type { MockRespFunc, ModuleMockHandler } from '@/types'

export const QUICK_MOCK_DATA_KEY = 'quickReqData'
export const REGEX_MOCK_DATA_KEY = 'restReqData'

const quickHandlerContext = new Map<string, ModuleMockHandler>()
const regexHandlerContext = new Array<ModuleMockHandler>()

export function transformRequest(
  req: Pick<IncomingMessage, 'url' | 'method'>,
) {
  // TODO
  // pick from cache
  // filter url and method from request, mapping to mockData
  const mockData = getMockHandler(req.url, req.method)

  if (mockData)
    wrapperHandler(mockData)

  return mockData
}

export type ResponseHandler = (_req: IncomingMessage, _res: ServerResponse) => (Object | void) | Promise<Object | void>

function wrapperHandler(handler: ModuleMockHandler) {
  const { response } = handler
  if (typeof response === 'object') {
    const originalRespData = response as Object
    handler.response = (_req, _res) => outputJson(_req, _res, originalRespData)
  }
  if (typeof response === 'function') {
    const originalRespFunc = response as MockRespFunc
    handler.response = async(_req, _res) => {
      // TODO
      // enhance _req, _res for mock
      const result = await originalRespFunc(_req, _res)
      if (result)
        outputJson(_req, _res, result)
    }
  }

  return handler
}

export function createTransformRequest(_config?: ResolvedConfig) {
  return async(_req: IncomingMessage,
    _res: ServerResponse,
    _next: (err?: any) => void,
  ) => {
    const handler = transformRequest(_req)
    if (handler) {
      const msg = `${colors.green('request hit')} ${colors.dim(_req.url)}`
      logger.info(msg)
      try {
        await (handler.response as Function)(_req, _res)
      }
      catch (error) {
        logger.error(colors.red((error as any).message))
      }
      return
    }
    _next()
  }
}

export function getMockHandler(url?: string, method?: string) {
  if (!url || !method)
    return null

  const handler = quickHandlerContext.get(`${method.toLowerCase()} ${url}`)
  if (handler)
    return handler

  for (const handler of regexHandlerContext) {
    const { url: _url, method: _method } = handler
    const matcher = parse(_url)
    if ((matcher.pattern.test(url) || handler.url === url)
      && _method.toLowerCase() === method.toLowerCase()
    )
      return handler
  }
  return null
}

export function setMockHandlerContext(values: ModuleMockHandler[], reset = true) {
  const logConflict = (source: ModuleMockHandler, target: ModuleMockHandler) => {
    logger.warn(`${colors.yellow('handler conflict as follows:')}
      ${colors.dim(colors.green(`${source.method.toUpperCase()} ${source.url}`))} ${colors.yellow(source._file)}
      ${colors.dim(colors.green(`${target.method.toUpperCase()} ${target.url}`))} ${colors.yellow(target._file)}`)
  }

  if (reset) {
    regexHandlerContext.splice(0, regexHandlerContext.length)
    quickHandlerContext.clear()
  }

  // restful priority > quick priority
  values
    .filter(handler => parse(handler.url).keys.length)
    .forEach((handler) => {
      const h = getMockHandler(handler.url, handler.method)
      if (h) {
        logConflict(handler, h)
        return
      }
      regexHandlerContext.push(handler)
    })
  values
    .filter(handler => !parse(handler.url).keys.length)
    .forEach((handler) => {
      const h = getMockHandler(handler.url, handler.method)
      if (h) {
        logConflict(handler, h)
        return
      }
      const key = `${handler.method.toLowerCase()} ${handler.url}`
      quickHandlerContext.set(key, handler)
    })
}

function outputJson(_req: IncomingMessage, res: ServerResponse, data: Object) {
  res.writeHead(200, { 'Content-Type': 'application/json' })
  res.end(JSON.stringify(data))
}
