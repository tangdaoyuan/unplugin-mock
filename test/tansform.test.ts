import { fileURLToPath } from 'url'
import { beforeAll, describe, expect, it } from 'vitest'
import { buildCjsFile, transformConfig } from '@/transform/config'
import type { MockHandler } from '@/types'
import { setMockData, transformRequest } from '@/transform/request'

const mockPath = fileURLToPath(new URL('./fixture', import.meta.url))

const pluginOptions = {
  mockPath,
  refresh: false,
  ignore: [],
}

describe('runs detect', () => {
  it('works', () => {
    const result = transformConfig(pluginOptions)
    expect(result).toBeInstanceOf(Array)
    expect(result.length).toBeGreaterThanOrEqual(1)
  })

  it('ignore file', () => {
    const result = transformConfig({ ...pluginOptions, ignore: ['**/ignore.ts'] })
    expect(result.some(r => r.url === '/api/ignore')).toBe(false)
  })
})

describe('runs file transform', () => {
  it('generate cjs for single mock file', () => {
    const code = buildCjsFile(fileURLToPath(new URL('./fixture/base.ts', import.meta.url)))
    expect(code).toMatchSnapshot()
  })
})

describe('runs mock request transform', () => {
  let mockList: MockHandler[] = []
  beforeAll(() => {
    mockList = transformConfig(pluginOptions)
    setMockData(mockList)
  })

  it('get json response', () => {
    const mockData = transformRequest({
      url: '/api/get',
      method: 'get',
    })
    expect(mockData).toMatchInlineSnapshot(`
      {
        "method": "get",
        "response": [Function],
        "url": "/api/get",
      }
    `)
  })
  it('post function response', () => {
    const mockData = transformRequest({
      url: '/api/post',
      method: 'post',
    })
    expect(mockData).toMatchInlineSnapshot(`
      {
        "method": "post",
        "response": [Function],
        "url": "/api/post",
      }
    `)
  })
  it('patch function response', () => {
    const mockData = transformRequest({
      url: '/api/post',
      method: 'post',
    })
    expect(mockData).toMatchInlineSnapshot(`
      {
        "method": "post",
        "response": [Function],
        "url": "/api/post",
      }
    `)
  })
  it('params function response', () => {
    const mockData = transformRequest({
      url: '/api/params/Tedy',
      method: 'get',
    })
    expect(mockData).toMatchInlineSnapshot(`
      {
        "method": "get",
        "response": [Function],
        "url": "/api/params/:name/:id?",
      }
    `)
    const multipleParamsData = transformRequest({
      url: '/api/params/Tedy/1',
      method: 'get',
    })
    expect(multipleParamsData).toMatchInlineSnapshot(`
      {
        "method": "get",
        "response": [Function],
        "url": "/api/params/:name/:id?",
      }
    `)
  })
})
