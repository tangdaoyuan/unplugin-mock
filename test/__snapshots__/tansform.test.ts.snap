// Vitest Snapshot v1

exports[`runs file transform > generate cjs for single mock file 1`] = `
"var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === \\"object\\" || typeof from === \\"function\\") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, \\"__esModule\\", { value: true }), mod);

// test/fixture/base.ts
var base_exports = {};
__export(base_exports, {
  default: () => base_default
});
module.exports = __toCommonJS(base_exports);
var base_default = [
  {
    url: \\"/api/base\\",
    method: \\"get\\",
    response: () => {
      return {
        code: 0,
        data: {
          name: \\"Tedy Base\\"
        }
      };
    }
  }
];
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {});
"
`;
